### ISSUE TITLE
User-declared options

### ISSUE THREAD

**Issue by avigad** [leanprover/lean3#1121]
It would be helpful to be able to declare user options in Lean, and have a tactic to get the value of an option. (Maybe these need to be `declare_option_nat`, `declare_option_bool`, ...)

**Comment by leodemoura** [comment]
Why do we need this feature?

In Lean 2, global options set with the command `set_option` were used to customize the behavior of built-in tactics. This approach has been abandoned in Lean 3 because a tactic is just a function and we can easily pass an arbitrary number of arguments to them. A global/user option is like a global variable. So, It seems a bad style to use them to control the behavior of tactics we implement in C++ or Lean. 
Note that, none of the existing built-in tactics do that in Lean 3.

We use options to control: pretty printer, trace messages, parser and elaborator.

@jroesch Do you want to add options for the native compiler using Lean? We can do this using C++. 

I don't see a strong motivation for this option. I'm planning to close this option.

**Comment by avigad** [comment]
Over the summer, we thought this would be a good way to set global options for automation, like
``` lean
set_option auto.use_classical true
set_option auto.max_steps 10000
```
I agree that this seems like bad style. It is o.k. with me to close.

But do you think it will be annoying to not be able to configure global behavior, e.g. to allow some automation to use `propext` or `funext` or classical logic in some setting? I suppose we can always make up a new tactic name for a classical version.

**Comment by leodemoura** [comment]
> I suppose we can always make up a new tactic name for a classical version.

Yes, we can do this.

We can use the new default structure values.

We can also use type classes:

```lean
class my_config :=
(val : nat) (flag : bool)

instance default_my_config : my_config :=
{val := 10, flag := tt}

open tactic my_config

meta def my_tac [my_config] : tactic unit :=
trace ("my_tac " ++ val^.to_string)

run_command my_tac -- my_tac 10

instance new_my_config : my_config :=
{val := 20, flag := ff}

run_command my_tac -- my_tac 20
```

**Comment by avigad** [comment]
Nice. Can we reliably assume that the most recent instance is the one that will be used?

I didn't know the default structure values had been implemented. I just found these:

https://github.com/leanprover/lean/commit/71a7a7f466ddb8e4fc1bf8edbe3a0e2154411a8b
https://github.com/leanprover/lean/commit/778d5382f612d85e63b34252404f9327e5504533
https://github.com/leanprover/lean/commit/6d12de633950dd96eecbb513996bdc9f70dfa3e9

This is very nice! :+1: 

These solutions are great. I will close this.
